registry/storage/driver/s3-aws/s3.go

```diff
@@ -1180,10 +1210,12 @@ func (d *driver) doWalk(parentCtx context.Context, objectCount *int64, path, pre
 				return false
 			}
 
-			if walkInfo.IsDir() {
-				if err := d.doWalk(ctx, objectCount, *walkInfo.prefix, prefix, f); err != nil {
-					retError = err
-					return false
+			if walkDirectories {
```

Comment:
nit: maybe we can combine this into 1 statement?

```go
if walkDirectories && walkInfo.IsDir() {

}
```

Reply:
what do we do `if !walkDirectories`?

Reply:
🤦🏻‍♂️ Good catch. The way this is implemented is a bit confusing. 

When `walkDirectories` is true then the `doWalk` function is a recursive depth first traversal, this condition controlling the recursive part. 

When `walkDirectories` is false, `doWalk` puts all of the traversal responsibility onto `ListObjectsV2PagesWithContext`, so there's no recursive component to it.